prow/cmd/deck/main.go

```diff
@@ -408,6 +435,201 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+		if src == "" {
+			http.Error(w, "missing src query parameter", http.StatusBadRequest)
+			return
+		}
+
+		page, err := renderSpyglass(sg, ca, src)
+		if err != nil {
+			logrus.WithError(err).Error("error rendering spyglass page")
+			http.Error(w, "error getting views for job", http.StatusInternalServerError)
+			return
+		}
+
+		fmt.Fprint(w, page)
+		elapsed := time.Since(start)
+		logrus.WithField("duration", elapsed).Info("Loading view from generic src completed.")
+	}
+}
+
+// renderSpyglass returns a pre-rendered Spyglass page from the given source string
+func renderSpyglass(sg *spyglass.Spyglass, ca *config.Agent, src string) (string, error) {
+	artifactNames, err := sg.ListArtifacts(src)
+	if err != nil {
+		return "", fmt.Errorf("error listing artifacts: %v", err)
+	}
+
+	viewerCache := map[string][]string{}
+	viewersRegistry := ca.Config().Deck.Spyglass.Viewers
+
+	for re, viewerNames := range viewersRegistry {
+		matches := []string{}
+		r, err := regexp.Compile(re)
+		if err != nil {
+			logrus.WithError(err).Error("Regexp failed to compile.")
+			continue
+		}
+		for _, a := range artifactNames {
+			if r.MatchString(a) {
+				matches = append(matches, a)
+			}
+		}
+		for _, vName := range viewerNames {
+			viewerCache[vName] = matches
+		}
+	}
+
+	viewsStart := time.Now()
+	lenses := sg.Views(viewerCache)
+	viewsElapsed := time.Since(viewsStart)
+	logrus.WithField("duration", viewsElapsed).Info("Loaded spyglass views")
+
+	var viewBuf bytes.Buffer
+	type ViewsTemplate struct {
+		Views       []spyglass.Lens
+		Source      string
+		ViewerCache map[string][]string
+	}
+	vTmpl := ViewsTemplate{
+		Views:       lenses,
+		Source:      src,
+		ViewerCache: viewerCache,
+	}
+	t, err := template.ParseFiles("/static/template/spyglass-template.html")
+	if err != nil {
+		return "", fmt.Errorf("error parsing template: %v", err)
+	}
+
+	if err = t.Execute(&viewBuf, vTmpl); err != nil {
+		return "", fmt.Errorf("error rendering template: %v", err)
+	}
+	return viewBuf.String(), nil
+}
+
+// handleArtifactView handles requests to load a single view for a job. This is what viewers
+// will use to call back to themselves.
+// Query params:
+// - name: required, specifies the name of the viewer to load
+// - src: required, specifies the job source from which to fetch artifacts
+func handleArtifactView(sg *spyglass.Spyglass, ca *config.Agent) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+		setHeadersNoCaching(w)
+		w.Header().Set("Content-Type", "application/json")
+		name := r.URL.Query().Get("name")
+		src := r.URL.Query().Get("src")
+		if name == "" {
+			http.Error(w, "missing name query parameter", http.StatusBadRequest)
+			return
+		}
+		if src == "" {
+			http.Error(w, "missing src query parameter", http.StatusBadRequest)
+			return
+		}
+
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			http.Error(w, "failed to read body", http.StatusBadRequest)
```

Comment:
I think failing to read the request body is a 5xx not a 4xx?

Reply:
I wasnt able to find that information, everywhere I looked said use generic 400 bad request, any source on that?

Reply:
Nope, it just seems like it should be a server error if we fail to read the request body. I guess the 4xx is assuming that the request body is malformed as a opposed to a network read error?

If 400 is typical lets keep that.

Reply:
Technically it's probably a 413 (payload too large) but we don't check what the error is so 400 makes sense